pytorch_lightning/callbacks/batch_size_finder.py

```diff
@@ -0,0 +1,425 @@
+from pytorch_lightning.utilities.parsing import lightning_getattr, lightning_hasattr, lightning_setattr
+from pytorch_lightning.utilities.rank_zero import rank_zero_info, rank_zero_warn
+
+if TYPE_CHECKING:
+    from pytorch_lightning.loggers.base import LightningLoggerBase
+
+    class _BatchSizeFinderDumpedParams(TypedDict):
+        callbacks: List[Callback]
+        logger: Optional[LightningLoggerBase]
+        max_steps: int
+        global_step: Optional[None]
+        limit_val_batches: Union[int, float]
+        limit_eval_batches: Union[int, float]
+
+
+log = logging.getLogger(__name__)
+
+
+class BatchSizeFinder(Callback):
+    SUPPORTED_MODES = ("power", "binsearch")
+
+    def __init__(
+        self,
+        mode: str = "power",
+        steps_per_trial: int = 3,
+        init_val: int = 2,
+        max_trials: int = 25,
+        batch_arg_name: str = "batch_size",
+    ) -> None:
+        """The `BatchSizeFinder` callback tries to find the largest batch size for a given model that does not give
+        an out of memory (OOM) error. It works with both training and evalation. All you need to do is add it as a
+        callback inside Trainer and call ``trainer.fit/validate/test/predict()``. Internally it calls the
+        respective step function ``steps_per_trial`` times for each batch size until one of the batch size
+        generates and OOM error.
+
+        Args:
+            mode: search strategy to update the batch size:
+
+                - ``'power'``: Keep multiplying the batch size by 2, until we get an OOM error.
+                - ``'binsearch'``: Initially keep multiplying by 2 and after encountering an OOM error
+                    do a binary search between the last successful batch size and the batch size that failed.
+
+            steps_per_trial: number of steps to run with a given batch size.
+                Ideally 1 should be enough to test if a OOM error occurs,
+                however in practice a few are needed.
+
+            init_val: initial batch size to start the search with.
+
+            max_trials: max number of increase in batch size done before
+               algorithm is terminated
+
+            batch_arg_name: name of the attribute that stores the batch size.
+                It is expected that the user has provided a model or datamodule that has a hyperparameter
+                with that name. We will look for this attribute name in the following places
+
+                - ``model``
+                - ``model.hparams``
+                - ``trainer.datamodule`` (the datamodule passed to the tune method)
+        """
+        mode = mode.lower()
+        if mode not in self.SUPPORTED_MODES:
+            raise MisconfigurationException(f"`mode` should be one of {self.SUPPORTED_MODES}")
+
+        self.mode = mode
+        self.steps_per_trial = steps_per_trial
+        self.init_val = init_val
+        self.max_trials = max_trials
+        self.batch_arg_name = batch_arg_name
```

Comment:
nit: preferably we should also do some input validation of these

Reply:
added a TODO for now.. will add them in a follow-up since we need to add tests for them too and if I add it here, the PR will grow more and will be hard for reviewers to review it properly.